File: ts/src/type.ts
// 基本的な鍵の種類の定義
export interface masterKey {
  keyType: "masterKeyPublic" | "masterKeyPrivate";
  key: string;
}

export interface identityKey {
  keyType: "identityKeyPublic" | "identityKeyPrivate";
  key: string;
  algorithm: string; // 例: "ML-DSA-65"
  timestamp: number;
  sessionUuid: string;
}

export interface accountKey {
  keyType: "accountKeyPublic" | "accountKeyPrivate";
  key: string;
  algorithm: string; // 例: "ML-KEM-768"
  timestamp: number;
}

export interface roomKey {
  keyType: "roomKey";
  key: string;
  algorithm: string; // 例: "AES-256-GCM"
  timestamp: number;
  sessionUuid: string;
}

export interface shareKey {
  keyType: "shareKeyPublic" | "shareKeyPrivate";
  key: string;
  algorithm: string; // 例: "ML-KEM-768"
  timestamp: number;
  sessionUuid: string;
}

export interface shareSignKey {
  keyType: "shareSignKeyPublic" | "shareSignKeyPrivate";
  key: string;
  algorithm: string; // 例: "ML-DSA-65"
  timestamp: number;
  sessionUuid: string;
}

export interface migrateKey {
  keyType: "migrateKeyPublic" | "migrateKeyPrivate";
  key: string;
  // 仕様書に基づきtimestampを追加
  timestamp?: number;
}

export interface migrateSignKey {
  keyType: "migrateSignKeyPublic" | "migrateSignKeyPrivate";
  key: string;
  // 仕様書に基づきtimestampを追加
  timestamp?: number;
}

export interface deviceKey {
  keyType: "deviceKey";
  key: string;
}

// 署名情報の定義
export interface Sign {
  keyHash: string;
  signature: string;
  keyType: string;
  algorithm?: string; // 署名アルゴリズム
}

// 暗号化データの定義
export interface EncryptedData {
  keyType: string;
  keyHash: string;
  encryptedData: string; // 仕様書ではbinaryEncryptedDataだが、コード側に合わせる
  iv: string; // 仕様書ではviだが、標準的な命名規則に合わせる
  algorithm?: string; // 暗号化アルゴリズム
  cipherText?: string;
}

// メッセージ関連の型定義
export type UserIdentifier = string;

export interface ReplyInfo {
  id: string;
}

// 各種コンテンツタイプの定義
export interface TextContent {
  text: string;
  format?: "plain" | "markdown";
  isThumbnail?: boolean;
  thumbnailOf?: string;
  originalSize?: number;
}

export interface ImageContent {
  uri: string;
  metadata: {
    filename: string;
    mimeType: string;
  };
  isThumbnail?: boolean;
  thumbnailOf?: string;
  originalSize?: number;
}

export interface VideoContent {
  uri: string;
  metadata: {
    filename: string;
    mimeType: string;
  };
  isThumbnail?: boolean;
  thumbnailOf?: string;
  originalSize?: number;
}

export interface AudioContent {
  uri: string;
  metadata: {
    filename: string;
    mimeType: string;
  };
  isThumbnail?: boolean;
  thumbnailOf?: string;
  originalSize?: number;
}

export interface FileContent {
  uri: string;
  metadata: {
    filename: string;
    mimeType: string;
  };
  isThumbnail?: boolean;
  thumbnailOf?: string;
  originalSize?: number;
}

// サムネイル関連の型定義
export interface TextThumbnail {
  originalType: "text";
  thumbnailText: string;
  size?: number;
}

export interface MediaThumbnail {
  originalType: "image" | "video";
  thumbnailUri: string;
  thumbnailMimeType: string;
  size?: number;
}

export interface FilesThumbnail {
  originalType: "file" | "audio";
  thumbnailText: string;
  size?: number;
}

export type ThumbnailContent = TextThumbnail | MediaThumbnail | FilesThumbnail;

export type MessageContent = TextContent | ImageContent | VideoContent | 
  AudioContent | FileContent | ThumbnailContent;

// メッセージの型定義（ドキュメントに合わせて更新）
export interface NotEncryptMessageValue {
  type: "text" | "image" | "video" | "audio" | "file" | "thumbnail";
  content: string; // 各タイプに応じたJSON文字列
  reply?: ReplyInfo; // ReplyInfo型を使用
  mention?: string[];
}

export interface NotEncryptMessage {
  encrypted: false;
  value: NotEncryptMessageValue;
  channel: string;
  original?: string;
  timestamp: number;
  isLarge: boolean;
  roomid: string; // roomidを追加
}

export interface EncryptedMessage {
  encrypted: true;
  value: string; // 暗号化されたデータ（EncryptedDataのJSON文字列）
  channel: string;
  original?: string;
  timestamp: number;
  isLarge: boolean;
  roomid: string; // roomidを追加
}

export type Message = NotEncryptMessage | EncryptedMessage;

// 既存の型定義は後方互換性のため残すか、削除するか検討
// export interface MessageContentType { ... }
// export interface MessageMetadataType { ... }
// export interface EncryptedMessageType { ... }

File: rust/src/type.rs
use serde::{Deserialize, Serialize};

/// 基本的な鍵の種類
#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct MasterKey {
    #[serde(rename = "keyType")]
    pub key_type: String,
    pub key: String,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct IdentityKey {
    #[serde(rename = "keyType")]
    pub key_type: String,
    pub key: String,
    pub algorithm: String,
    pub timestamp: u64,
    #[serde(rename = "sessionUuid")]
    pub session_uuid: String,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct AccountKey {
    #[serde(rename = "keyType")]
    pub key_type: String,
    pub key: String,
    pub algorithm: String,
    pub timestamp: u64,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct ServerKey {
    #[serde(rename = "keyType")]
    pub key_type: String,
    pub key: String,
    pub timestamp: u64,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct RoomKey {
    #[serde(rename = "keyType")]
    pub key_type: String,
    pub key: String,
    pub algorithm: String,
    pub timestamp: u64,
    #[serde(rename = "sessionUuid")]
    pub session_uuid: String,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct ShareKey {
    #[serde(rename = "keyType")]
    pub key_type: String,
    pub key: String,
    pub algorithm: String,
    pub timestamp: u64,
    #[serde(rename = "sessionUuid")]
    pub session_uuid: String,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct ShareSignKey {
    #[serde(rename = "keyType")]
    pub key_type: String,
    pub key: String,
    pub algorithm: String,
    pub timestamp: u64,
    #[serde(rename = "sessionUuid")]
    pub session_uuid: String,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct MigrateKey {
    #[serde(rename = "keyType")]
    pub key_type: String,
    pub key: String,
    pub timestamp: Option<u64>,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct MigrateSignKey {
    #[serde(rename = "keyType")]
    pub key_type: String,
    pub key: String,
    pub timestamp: Option<u64>,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct DeviceKey {
    #[serde(rename = "keyType")]
    pub key_type: String,
    pub key: String,
}

/// 署名情報
#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct Sign {
    #[serde(rename = "keyHash")]
    pub key_hash: String,
    pub signature: String,
    #[serde(rename = "keyType")]
    pub key_type: String,
    pub algorithm: Option<String>,
}

/// 暗号化データ
#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct EncryptedData {
    #[serde(rename = "keyType")]
    pub key_type: String,
    #[serde(rename = "keyHash")]
    pub key_hash: String,
    #[serde(rename = "encryptedData")]
    pub encrypted_data: String,
    pub iv: String,
    pub algorithm: Option<String>,
    #[serde(rename = "cipherText")]
    pub cipher_text: Option<String>,
}

pub type UserIdentifier = String;

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct ReplyInfo {
    pub id: String,
}

/// 各種コンテンツ
#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct TextContent {
    pub text: String,
    pub format: Option<String>,
    #[serde(rename = "isThumbnail")]
    pub is_thumbnail: Option<bool>,
    #[serde(rename = "thumbnailOf")]
    pub thumbnail_of: Option<String>,
    #[serde(rename = "originalSize")]
    pub original_size: Option<u64>,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct MediaMetadata {
    pub filename: String,
    #[serde(rename = "mimeType")]
    pub mime_type: String,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct ImageContent {
    pub uri: String,
    pub metadata: MediaMetadata,
    #[serde(rename = "isThumbnail")]
    pub is_thumbnail: Option<bool>,
    #[serde(rename = "thumbnailOf")]
    pub thumbnail_of: Option<String>,
    #[serde(rename = "originalSize")]
    pub original_size: Option<u64>,
}

pub type VideoContent = ImageContent;
pub type AudioContent = ImageContent;
pub type FileContent = ImageContent;

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(tag = "originalType")]
pub enum ThumbnailContent {
    #[serde(rename = "text")]
    Text {
        #[serde(rename = "thumbnailText")]
        thumbnail_text: String,
        size: Option<u64>,
    },
    #[serde(rename = "image")]
    Image {
        #[serde(rename = "thumbnailUri")]
        thumbnail_uri: String,
        #[serde(rename = "thumbnailMimeType")]
        thumbnail_mime_type: String,
        size: Option<u64>,
    },
    #[serde(rename = "video")]
    Video {
        #[serde(rename = "thumbnailUri")]
        thumbnail_uri: String,
        #[serde(rename = "thumbnailMimeType")]
        thumbnail_mime_type: String,
        size: Option<u64>,
    },
    #[serde(rename = "file")]
    File {
        #[serde(rename = "thumbnailText")]
        thumbnail_text: String,
        size: Option<u64>,
    },
    #[serde(rename = "audio")]
    Audio {
        #[serde(rename = "thumbnailText")]
        thumbnail_text: String,
        size: Option<u64>,
    },
}

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(untagged)]
pub enum MessageContent {
    Text(TextContent),
    Image(ImageContent),
    Video(VideoContent),
    Audio(AudioContent),
    File(FileContent),
    Thumbnail(ThumbnailContent),
}

/// メッセージ本体
#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct NotEncryptMessageValue {
    #[serde(rename = "type")]
    pub _type: String,
    pub content: String,
    pub reply: Option<ReplyInfo>,
    pub mention: Option<Vec<String>>,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct NotEncryptMessage {
    pub encrypted: bool,
    pub value: NotEncryptMessageValue,
    pub channel: String,
    pub original: Option<String>,
    pub timestamp: u64,
    #[serde(rename = "isLarge")]
    pub is_large: bool,
    pub roomid: String,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct EncryptedMessage {
    pub encrypted: bool,
    pub value: String,
    pub channel: String,
    pub original: Option<String>,
    pub timestamp: u64,
    #[serde(rename = "isLarge")]
    pub is_large: bool,
    pub roomid: String,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(untagged)]
pub enum Message {
    NotEncrypted(NotEncryptMessage),
    Encrypted(EncryptedMessage),
}
